src/main/java/net/neoforged/neoforge/attachment/AttachmentType.java

```diff
@@ -105,118 +99,92 @@ public static <T> Builder<T> builder(Supplier<T> defaultValueSupplier) {
      *
      * @param defaultValueConstructor A constructor for a new default value of this attachment type.
      */
-    public static <T> Builder<T> builder(Function<IAttachmentHolder, T> defaultValueConstructor) {
-        return new Builder<>(defaultValueConstructor);
+    public static <T> Builder<T, IAttachmentHolder> builder(Function<IAttachmentHolder, T> defaultValueConstructor) {
+        return Builder.fromReflection(defaultValueConstructor);
     }
 
-    /**
-     * Create a builder for an attachment type that uses {@link INBTSerializable} for serialization.
-     * Other kinds of serialization can be implemented using {@link #builder(Supplier)} and {@link Builder#serialize(IAttachmentSerializer)}.
-     *
-     * <p>See {@link #serializable(Function)} for attachments that want to capture a reference to their holder.
-     */
-    public static <S extends Tag, T extends INBTSerializable<S>> Builder<T> serializable(Supplier<T> defaultValueSupplier) {
-        return serializable(holder -> defaultValueSupplier.get());
+    public static <T, P extends IAttachmentHolder> Builder<T, P> builder(Class<P> holderClass, Function<P, T> defaultValueConstructor) {
+        return Builder.fromReflection(holderClass, defaultValueConstructor);
     }
 
-    /**
-     * Create a builder for an attachment type that uses {@link INBTSerializable} for serialization.
-     * Other kinds of serialization can be implemented using {@link #builder(Supplier)} and {@link Builder#serialize(IAttachmentSerializer)}.
-     *
-     * <p>This overload allows capturing a reference to the {@link IAttachmentHolder} for the attachment.
-     * To obtain a specific subtype, the holder can be cast.
-     * If the holder is of the wrong type, the constructor should throw an exception.
-     * See {@link #serializable(Supplier)} for an overload that does not capture the holder.
-     */
-    public static <S extends Tag, T extends INBTSerializable<S>> Builder<T> serializable(Function<IAttachmentHolder, T> defaultValueConstructor) {
-        return builder(defaultValueConstructor).serialize(new IAttachmentSerializer<S, T>() {
-            @Override
-            public T read(IAttachmentHolder holder, S tag, HolderLookup.Provider provider) {
-                var ret = defaultValueConstructor.apply(holder);
-                ret.deserializeNBT(provider, tag);
-                return ret;
-            }
-
-            @Nullable
-            @Override
-            public S write(T attachment, HolderLookup.Provider provider) {
-                return attachment.serializeNBT(provider);
-            }
-        });
-    }
-
-    public static class Builder<T> {
-        private final Function<IAttachmentHolder, T> defaultValueSupplier;
+    public static class Builder<T, P extends IAttachmentHolder> {
+        private final Function<P, T> defaultValueSupplier;
+        private final Class<P> attachmentHolderClass;
```

Comment:
This also doesn't seem to be used anywhere unless I'm missing something? And once again, you go to a lot of hacky trouble to get it.